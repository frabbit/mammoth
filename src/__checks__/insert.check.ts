import {
  count,
  defineDb,
  defineTable,
  integer,
  text,
  timestampWithTimeZone,
  uuid,
} from '../../.build';

import { Query } from '../../.build/query';
import { ResultSet } from '../../.build/result-set';

const toSnap = <T extends Query<any>>(query: T): ResultSet<T, true> => {
  return undefined as any;
};

/** @dts-jest enable:test-type */

const foo = defineTable({
  id: uuid().primaryKey().default(`gen_random_uuid()`),
  createDate: timestampWithTimeZone().notNull().default(`now()`),
  name: text().notNull(),
  value: integer(),
});

const db = defineDb({ foo }, () => Promise.resolve({ rows: [], affectedCount: 0 }));

// @dts-jest:group insert check
{
  // @dts-jest:snap should insert and returning count
  toSnap(db.insertInto(db.foo).values({ name: `Test` }));

  // @dts-jest:snap should insert multiple rows and returning count
  toSnap(db.insertInto(db.foo).values([{ name: `Test` }, { name: `Test 2` }]));

  // @dts-jest:snap should insert default column
  toSnap(db.insertInto(db.foo).values({ name: `Test`, createDate: new Date() }));

  // @dts-jest:fail:snap should not insert unknown column
  toSnap(db.insertInto(db.foo).values({ name: `Test`, asd: `Test` }));

  // @dts-jest:fail:snap should not insert invalid type in known column
  toSnap(db.insertInto(db.foo).values({ name: 123 }));

  // @dts-jest:fail:snap should not insert multiple rows with invalid colums
  toSnap(db.insertInto(db.foo).values([{ name: `Test` }, { name: `Test 2`, asd: 123 }]));

  db.insertInto(db.foo)
    .values({ name: `Test` })
    .then((result) => {
      // @dts-jest:snap should insert and await affect count
      result;
    });

  db.insertInto(db.foo)
    .values({ name: `Test` })
    .returning(`name`)
    .then((result) => {
      // @dts-jest:snap should insert-returning and await rows
      result;
    });
}
