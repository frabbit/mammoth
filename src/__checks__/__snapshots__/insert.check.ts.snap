// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`insert check should insert and await affect count (type) should match snapshot 1`] = `"number"`;

exports[`insert check should insert and returning count (type) should match snapshot 1`] = `"number"`;

exports[`insert check should insert default column (type) should match snapshot 1`] = `"number"`;

exports[`insert check should insert multiple rows and returning count (type) should match snapshot 1`] = `"number"`;

exports[`insert check should insert-returning and await rows (type) should match snapshot 1`] = `"{ name: string; }[]"`;

exports[`insert check should not insert invalid type in known column (type) should match snapshot 1`] = `"Type 'number' is not assignable to type 'string'."`;

exports[`insert check should not insert multiple rows with invalid colums (type) should match snapshot 1`] = `
"Type '{ name: string; asd: number; }' is not assignable to type '{ name: string; } & { id?: string; createDate?: Date; name?: string; value?: number; }'.
  Object literal may only specify known properties, and 'asd' does not exist in type '{ name: string; } & { id?: string; createDate?: Date; name?: string; value?: number; }'."
`;

exports[`insert check should not insert unknown column (type) should match snapshot 1`] = `
"Argument of type '{ name: string; asd: string; }' is not assignable to parameter of type '({ name: string; } & { id?: string; createDate?: Date; name?: string; value?: number; }) | ({ name: string; } & { id?: string; createDate?: Date; name?: string; value?: number; })[]'.
  Object literal may only specify known properties, and 'asd' does not exist in type '({ name: string; } & { id?: string; createDate?: Date; name?: string; value?: number; }) | ({ name: string; } & { id?: string; createDate?: Date; name?: string; value?: number; })[]'."
`;
